lib/csv/decoder.ex

```diff
@@ -61,120 +61,102 @@ defmodule CSV.Decoder do
+    decode(stream, options)
+    |> handle_errors!
+  end
 
+  def decode(stream, options \\ []) do
+    { stream, options |> options_with_defaults }
+    |> prepare_headers
+    |> prepare_row_length
+    |> decode_stream
+  end
 
+  defp decode_stream({ stream, %{ multiline_escape: multiline_escape } = options }) do
     stream
     |> aggregate(multiline_escape)
     |> Stream.with_index
-    |> ParallelStream.map(fn { line, index } ->
-      { line, index }
-      |> Lexer.lex(options)
-      |> lex_to_parse
-      |> Parser.parse(options)
-      |> check_row_length(row_length)
-      |> build_row(headers)
-    end, options |> Keyword.merge(num_workers: num_workers))
-    |> handle_errors
-  end
-  defp aggregate(stream, true) do
-    stream |> LineAggregator.aggregate
-  end
-  defp aggregate(stream, false) do
-    stream
+    |> ParallelStream.map(&(decode_row(&1, options)),
+                          options |> Enum.into([]))
   end
 
-  defp lex_to_parse({ :ok, tokens, index }) do
-    { tokens, index }
-  end
-  defp lex_to_parse(result) do
-    result
+  defp decode_row({ nil, 0 }, %{ row_length: false }), do: { :ok, [] }
+  defp decode_row({ line, index }, %{ headers: headers, row_length: row_length } = options) do
+    with { :ok, parsed, _ } <- parse_row({ line, index }, options |> Enum.into([])),
+         { :ok } <- check_row_length({ parsed, index }, row_length),
+         do: build_row(parsed, headers)
   end
 
-  defp check_row_length({ :ok, data, index }, row_length) do
-    actual_length = data |> Enum.count
-
-    case actual_length do
-      ^row_length -> { :ok, data, index }
-      _ -> { :error, RowLengthError, "Encountered a row with length #{actual_length} instead of #{row_length}", index }
-    end
-  end
-  defp check_row_length(error, _) do
-    error
+  defp parse_row({ line, index}, options) do
+    with { :ok, lex, _ } <- Lexer.lex({ line, index }, options),
+    do: Parser.parse({ lex, index }, options)
   end
 
-  defp build_row({ :ok, data, _ }, headers) when is_list(headers) do
-    headers |> Enum.zip(data) |> Enum.into(%{})
+  defp aggregate(stream, true) do
+    stream |> LineAggregator.aggregate
   end
-  defp build_row({ :ok, data, _ }, _) do
-    data
+  defp aggregate(stream, false), do: stream
+
+  defp build_row(data, headers) when is_list(headers) do
+    { :ok, headers |> Enum.zip(data) |> Enum.into(%{}) }
   end
-  defp build_row(error, _) do
-    error
+  defp build_row(data, _), do: { :ok, data }
+
+  defp options_with_defaults(options) do
+    num_pipes = options |> Keyword.get(:num_pipes, Defaults.num_workers)
+    options
+    |> Keyword.merge(num_pipes: num_pipes,
+                     num_workers: options |> Keyword.get(:num_workers, num_pipes),
+                     multiline_escape: options |> Keyword.get(:multiline_escape, true),
+                     headers: options |> Keyword.get(:headers, false),
+                     row_length: false)
+    |> Enum.into(%{})
```

Comment:
As you already mentioned, this has some downstream problems.

From the top of my head, we could:
1. Use an options module to make the intent more clear (`options |> Options.as_map` etc) but that would probably be largely cosmetic
2. Use a struct that has the original keyword list as a field
3. Revert to using `Keyword.get` for the time being

After some thought I'd favor option 3. I appreciate that using a map would make for better internal lib code. My main concern is not getting this pr sidetracked - this could be solved in another pr looking at the other module's APIs